pkg/testutil/test/test.go

```diff
@@ -0,0 +1,118 @@
+// A Command can be used as a Case Command obviously, but also as part of a Setup or Cleanup routine,
+// and as the basis of any type of helper.
+// A Command can be cloned, in which case, the subcommand inherits a copy of all of its Env and parameters.
+// Typically, a Case has a base-command, from which all commands involved in the test are derived.
+type Command interface {
+	// WithBinary specifies what binary to execute
+	WithBinary(binary string)
+	// WithArgs specifies the args to pass to the binary. Note that WithArgs is additive.
+	WithArgs(args ...string)
+	// WithEnv adds the passed map to the environment of the command to be executed
+	WithEnv(env map[string]string)
+	// WithWrapper allows wrapping a command with another command (for example: `time`, `unbuffer`)
+	WithWrapper(binary string, args ...string)
+	// Run does execute the command, and compare the output with the provided expectation.
+	// Passing nil for `Expected` will just run the command regardless of outcome.
+	// An empty `&Expected{}` is (of course) equivalent to &Expected{Exit: 0}, meaning the command is verified to be
+	// successful
+	Run(expect *Expected)
+	// Clone returns a copy of the command
+	Clone() Command
+	// Clear will clear binary and arguments, but retain the env, or any other custom properties
+	Clear()
+}
+
+type Comparator func(stdout string, info string, t *testing.T)
+
+// Expected expresses the expected output of a command
+type Expected struct {
+	// ExitCode to expect
+	ExitCode int
+	// Errors contains any error that (once serialized) should be seen in stderr
+	Errors []error
+	// Output function to match against stdout
+	Output Comparator
+}
+
+type ConfigKey string
+type ConfigValue string
+
+type SystemKey string
+type SystemValue string
+
+// Data is meant to hold information about a test:
+// - first, any random key value data that the test implementer wants to carry / modify - this is test data
+// - second, configuration specific to the binary being tested - typically defined by the specialized command being tested
+// - third, immutable "system" info (unique identifier, tempdir, or other SystemKey/Value pairs)
+type Data interface {
+	// Get returns the value of a certain key for custom data
+	Get(key string) string
+	// Set will save `value` for `key`
+	Set(key string, value string) Data
+
+	// Identifier returns the test identifier that can be used to name resources
+	Identifier() string
+	// TempDir returns the test temporary directory
+	TempDir() string
+	// Sink allows to define ONCE a certain system property
+	Sink(key SystemKey, value SystemValue)
+	// Surface allows retrieving a certain system property
+	Surface(key SystemKey) SystemValue
+
+	// WithConfig allows setting a declared ConfigKey to a ConfigValue
+	WithConfig(key ConfigKey, value ConfigValue) Data
+	ReadConfig(key ConfigKey) ConfigValue
+
+	// Private methods
```

Comment:
private methods do not need to be defined in the interface?

Reply:
I do need getLabels and getConfig here: 
https://github.com/containerd/nerdctl/blob/f1434f42d18ca848b7d84d36c13bf073ab2f6827/pkg/testutil/test/data.go#L131-L137

Will remove `adopt`, which is not necessary.